diff --git a/third_party/ffmpeg/chromium/config/Chrome/win/ia32/config.h b/third_party/ffmpeg/chromium/config/Chrome/win/ia32/config.h
index af12de7dff..fe47509bb7 100644
--- a/third_party/ffmpeg/chromium/config/Chrome/win/ia32/config.h
+++ b/third_party/ffmpeg/chromium/config/Chrome/win/ia32/config.h
@@ -2240,7 +2240,7 @@
 #define CONFIG_DATA_PROTOCOL 0
 #define CONFIG_FFRTMPCRYPT_PROTOCOL 0
 #define CONFIG_FFRTMPHTTP_PROTOCOL 0
-#define CONFIG_FILE_PROTOCOL 0
+#define CONFIG_FILE_PROTOCOL 1
 #define CONFIG_FTP_PROTOCOL 0
 #define CONFIG_GOPHER_PROTOCOL 0
 #define CONFIG_HLS_PROTOCOL 0
diff --git a/third_party/ffmpeg/chromium/config/Chrome/win/ia32/libavformat/protocol_list.c b/third_party/ffmpeg/chromium/config/Chrome/win/ia32/libavformat/protocol_list.c
index 3155bcfc19..4efbda5db2 100644
--- a/third_party/ffmpeg/chromium/config/Chrome/win/ia32/libavformat/protocol_list.c
+++ b/third_party/ffmpeg/chromium/config/Chrome/win/ia32/libavformat/protocol_list.c
@@ -1,2 +1,3 @@
 static const URLProtocol *url_protocols[] = {
+    &ff_file_protocol,
     NULL };
diff --git a/third_party/ffmpeg/ffmpeg_generated.gni b/third_party/ffmpeg/ffmpeg_generated.gni
index a5aa3a4623..9a498be608 100644
--- a/third_party/ffmpeg/ffmpeg_generated.gni
+++ b/third_party/ffmpeg/ffmpeg_generated.gni
@@ -104,6 +104,7 @@ if ((is_android && current_cpu == "arm" && arm_use_neon) || (is_android && curre
     "libavformat/oggparsevp8.c",
     "libavformat/os_support.c",
     "libavformat/protocols.c",
+    "libavformat/file.c",
     "libavformat/qtpalette.c",
     "libavformat/rawdec.c",
     "libavformat/replaygain.c",
